### DISCUSSION TITLE
Typescript issue with 5.0.2

### DISCUSSION THREAD

**Discussion by ZebraFlesh** [pmndrs/zustand#2926] [Bug report]
### Bug Description

I think #2875 created a Typescript issue. With 5.0.1, this code is perfectly fine:

```
export const createStore = <TStore extends Record<string, unknown>>(
  getStore: (set: NamedSet<TStore>) => TStore,
) => create<TStore>()(devtools(getStore, { enabled: process.env.NODE_ENV === 'development' }));
```

But with 5.0.2, I get the following error:

```
error TS2345: Argument of type 'StateCreator<TStore | Partial<TStore>, [], [["zustand/devtools", never]]>' is not assignable to parameter of type 'StateCreator<TStore, [], [["zustand/devtools", never]]>'.
  Type 'StateCreator<TStore | Partial<TStore>, [], [["zustand/devtools", never]]>' is not assignable to type '(setState: { (partial: TStore | Partial<TStore> | ((state: TStore) => TStore | Partial<TStore>), replace?: false | undefined): void; (state: TStore | ((state: TStore) => TStore), replace: true): void; }, getState: () => TStore, store: StoreApi<...>) => TStore'.
    Type 'TStore | Partial<TStore>' is not assignable to type 'TStore'.
      'TStore | Partial<TStore>' is assignable to the constraint of type 'TStore', but 'TStore' could be instantiated with a different subtype of constraint 'Record<string, unknown>'.
        Type 'Partial<TStore>' is not assignable to type 'TStore'.
          'Partial<TStore>' is assignable to the constraint of type 'TStore', but 'TStore' could be instantiated with a different subtype of constraint 'Record<string, unknown>'.

   9 | export const createStore = <TStore extends Record<string, unknown>>(
  10 |   getStore: (set: NamedSet<TStore>) => TStore,
> 11 | ) => create<TStore>()(devtools(getStore, { enabled: process.env.NODE_ENV === 'development' }));
     |                       ^
  12 | 
```

(I would create a reproduction, but I've inherited this project and am not proficient enough with Zustand to do this without a stackblitz template or similar.)

For now, downgrading to 5.0.1 is the answer.

### Reproduction Link

see above

**Comment by dai-shi** [comment]
@AndyRightNow Can you help this?

**Reply by AndyRightNow** [reply]
> @AndyRightNow Can you help this?

@dai-shi Sure, let me check

**Comment by AndyRightNow** [comment]
I just created a minimal reproduction here https://stackblitz.com/edit/vitejs-vite-b4fksmbb?file=src%2Fstore.ts

However it's weird that I can't seem to be able to reproduce it locally in unit tests. I will look into it and find a fix when I can.

**Reply by dai-shi** [reply]
just curious if it can be reproduced with typescript playground.

**Reply by AndyRightNow** [reply]
@dai-shi I tried and it's the same as reproducing in unit tests. No errors 🤔

[TS Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzQGjjATzDTgGUYs0BhdLaOAXzgDMoIQ4AiALwFcBnCgDsAJpwDcAKFCRYiOCLQA3GBAgAbfnkLE4AOQwg0I0mnhNW7Ln0EZRAehDARI9WgDuGdBMmS0AD1l4ZAghQRRabHJoEgBeOAAeABVo9DgA7FF+OAAlNBCoEQTBKGAhAHM8XiEAayEIdyEAPiaACkk4OHKzVLQALjhW-jMBgyMTM2TepoBKOFimuBTVdEk5hYjMbCmVtDaZ1sUVNU1W7phevCQ0IQwAIzcRAeYMTRIGGZmpP0DoYNDwqgtmgAOrQGpeCDVEy7eaJZYxdL+TIibJ5ApFEplSpwap1BrNNodLo9XYDciUGhbaA7GJNNbzRZAyi09D7Q7KVQafhnUkxK7pW4PYzPV7DRifKRAA)

**Comment by AndyRightNow** [comment]
@ZebraFlesh 

Could you upgrade to `5.0.2` and use the following workaround for now?

```diff
export const createStore = <TStore extends Record<string, unknown>>(
-  getStore: (set: NamedSet<TStore>) => TStore,
+  getStore: StateCreator<TStore>,
) => create<TStore>()(devtools(getStore, { enabled: process.env.NODE_ENV === 'development' }));
```

Not sure why the relatively low-level type `NamedSet` is used in your implementation but after https://github.com/pmndrs/zustand/pull/2875, the return type has been factored in so it'd be better to use the complete type as well.

I also added this workaround to the demo https://stackblitz.com/edit/vitejs-vite-b4fksmbb?file=src%2Fstore.ts

**Reply by dai-shi** [reply]
Oh, yeah, I think we may want to consider `NamedSet` as an internal type.

**Reply by ZebraFlesh** [reply]
Thanks, that works for me!
